title: "Phosphorus Traits"
author: "Jay T. Lennon, Mario Muscarella, Kali Bird"
date: "`r format(Sys.time(), '%d %B, %Y')`"
header-includes:
  - \usepackage{array}
output: pdf_document
geometry: margin=2.54cm
---

Generalist and specialist strategies of phosphorus acquistion by aquatic bacteria 

```{r}
rm(list=ls())
getwd()
setwd("~/GitHub/PhosTrait")
```

Load package for writing figures
```{r}
require("png")
require("vegan")
require("phylolm")
require("muscle")
require("seqinr")
require("ape")
```

Load data
```{r}
gr.data <- read.csv("data/grraw.csv", sep=",", header = TRUE, row.names = 1)
str(gr.data)

# Standadize Growth Rates Across Strains
gr.std <- gr.data / (apply(gr.data, 1, sum))

# Calculate Max Growth Rate
umax <- as.matrix((apply(gr.data, 1, max))) 
```

Niche breadth (Levins 1968)
```{r}
levins <- function(p_xi = ""){
  p = 0
  for (i in p_xi){
    p = p + i^2
    }
  nb = 1 / (length(p_xi) * p)
  return(nb)
  }

# Calculate niche breadth for Each Isolate
nb <- as.matrix(levins(gr.std))

# Add Row & Column Names to Niche Breadth Matrix
rownames(nb) <- rownames(gr.data)
colnames(nb) <- c("nb")
```  

Build a `phylo` file
```{r}
# Identify input and output files to perform alignment in the `muscle` package
muscle::muscle(seqs="./data/p.isolates.fasta", out = "./data/p.isolates.afa") 

# Read Alignment File {seqinr}
read.aln <- read.alignment(file = "./data/p.isolates.afa", format = "fasta")  

# Convert Alignment File to DNAbin Object {ape}
p.DNAbin <- as.DNAbin(read.aln) 

# Identify Base Pair Region of 16S rRNA Gene to Visuzlize
window <- p.DNAbin[, 100:500] 

# Command to Visusalize Sequence Alignment {ape}
image.DNAbin(window, cex.lab = 0.50) 

# Create Distance Matrix with "F84" Model {ape}
seq.dist.84 <- dist.dna(p.DNAbin, model = "F84", pairwise.deletion = FALSE)

# Neighbor Joining Algorithm to Construct Tree, a 'phylo' Object {ape}
nj.tree <- bionj(seq.dist.84)

# Identify Outgroup Sequence
outgroup <- match("Methanosarcina", nj.tree$tip.label)

# Root the Tree {ape}
nj.rooted <- root(nj.tree, outgroup, resolve.root = TRUE)

# Plot the Rooted Tree{ape}
par(mar = c(1,1,2,1) + 0.1)
plot.phylo(nj.rooted, main = "Neigbor Joining Tree", "phylogram", 
           use.edge.length = FALSE, direction = "right", cex = 0.6, 
           label.offset = 1)
add.scale.bar(cex = 0.7)

# Keep Rooted but Drop Outgroup Branch
p.tree <- drop.tip(nj.rooted, "Methanosarcina")
```

Phylogenetic regression (simple w/o dummy variable)
```{r}
taxa <- sort(p.tree$tip.label)
dat <- data.frame(umax, nb)
fit <- phylolm(umax ~ nb, data = dat, phy = p.tree, model = "lambda")
summary(fit)
```

Phylogenetic multiple regression
```{r}
lake <- ifelse(grepl("WG",row.names(gr.data)),'WG', 'LL') # make an empty vector for lake id
tradeoff <- data.frame(nb, umax, lake) # make new data frame
D <- (lake == "LL") * 1
fit <- phylolm(log10(tradeoff$umax) ~ tradeoff$nb + D + tradeoff$nb * D, data = tradeoff, 
        phy = p.tree, model = "lambda")
summary(fit)
```

# EXAMPLE FROM PHYLOLM THAT SEEMS TO WORK JUST FINE
# https://cran.r-project.org/web/packages/phylolm/phylolm.pdf
set.seed(123456)
tre = rcoal(60)
taxa = sort(tre$tip.label)
b0=0; b1=1;
x <- rTrait(n=1, phy=tre,model="BM",
parameters=list(ancestral.state=0,sigma2=10))
y <- b0 + b1*x +
rTrait(n=1,phy=tre,model="lambda",parameters=list(
ancestral.state=0,sigma2=1,lambda=0.5))
dat = data.frame(trait=y[taxa],pred=x[taxa])
fit = phylolm(trait~pred,data=dat,phy=tre,model="lambda")
summary(fit)

